### DISCUSSION TITLE
Allow "positive" airspaces / improve airspace handling a lot

### DISCUSSION THREAD

**Discussion by groundhogxc** [XCSoar/XCSoar#1468] [Ideas]
I would like to suggest changes in the management of airspaces that could be implemented, whether this is the best ways should be open for discussion. I did not find a previous issue about this, please point me if I overlooked it.

Current situation:
===========

The idea is to improve the following three aspects:

1) *Acknowledgements*: Currently, when entering an airspace which is not prohibited (either not active or due to ATC clearance), there are only two options to proceed: "Acknowledge", and "Acknowledge Day". The first makes the warning go away for a very short time before it returns, the other disables the warning for the entire day. In practice, often neither of these are what you want - e.g. you want to cross the airspace once, and then it should ideally be automatically reactivated (see below)

2) *Clearance altitudes*: When cleared by ATC into an airspace, often altitude limits are imposed. The controller will clear you into the airspace "up to flight level 100" or so, Or, a restriction area might be activated on a given day, but it sometimes is a different altitude up to which it is active on different days. I encounter both of these cases very often. So, there should be a way to not completely the warnings for a given airspace, but rather to (temporarily) adjust its altitude limits.

3) *"Positive" airspaces*: There are many airspaces which are not restrictions, but rather excepetions from restrictions of the bigger restrictive airspace they are inside of (when active): glider sectors and wave sectors, typically. Currently these either generate warnings, or are ignored; but really they should "disable" warnings as long as one is *inside* the sector (basically an inverted behaviour). 

Proposed changes / implementation:
=======================

General:  Personally I do not like the word "acknowledge" here, but maybe there is a good reason for it. I think it would be better to use something like "clearance" - but maybe there are good reasons against it. Nonetheless, for this reason, and to distinguish from the current situation, I will use "clearance" here.
Depending on how powerful we want the mechanism to be, we might need two states for an airspace: Enabled / Disabled and Cleared/Not cleared (see below)

Here are my ideas relevant to the points 1-3 above;

1) There should be at least one additional option "Clearance - Once", which should suppress the airspace warnings for that airspace until the aircraft has left the airspace for some amoiunt of time (say one minute, to avoid problems when circling on the edge of the airspace - where, as we all know, usually the best thermals live). There could also be specific timers associated such as "Clearance - 30 minutes" - not sure how desired that would be.

2) There should be an option "clearance up to XXXX" with a way to enter an altitude and "clearance above XXXX" similarly. This should keep the airspace active in the normal way, but modify, temporarily, the altitude limits. Again it should have tim eoptions such as "Until exit", "Now (like, 1 minute)", "All day" etc. as discussed above. (In Switzerland, that would also an alternative way to deal with "MIL ON"/"MIL OFF" without maintaining a separate airspace file)

3) For glider sectors, etc., so "positive airspaces", I propose that airspace warnings should be suppressed, as long as ONE of the airspaces that the aircraft currently is in (or even: is about to enter) has a valid "clearance" set.  Ideally, this should take into account warnings when you are about to LEAVE that permitted airspace (which is a more complex addition), as well as the modified altitude limits from (2). This should cover the gliding sector situation nicely.

However: The "clearance" approach from 3) generates possible problems: The typical situation is a restriction area, say, from GND to a high altitude limit, which is partially inside another airspace (typically some restricted airspace such as CHARLIE starting at some FL which is lower than the top FL of the restriction), and is currently NOT active. If you set this to "clearance-day", then it would suppress the warnings when you as ascending into the CHARLIE inside the inactive restricted zone, wich would be bad. Theoretically, this can be handled by setting a clearance only up to the lower level of the CHARLIE, is (2) is implemented, but that would require actively figuring out the local airspace structure. Also does not work if that altitude limit is different in different parts of the restricted area.

So, it would be much better if there is a separate state "Enabled/Disabled" (which is better than "Active"/"Not active", see below), which deactivates that airspace, but does not disable warnings for other airspaces inside it's volume. It should ideally even use the same time structure (disable temporarily, disable once, disable day) as the "clearance" state; in fact I think maybe one single timer should be enough for both states (in both cases after the "until" criterion expires, the airspace should return to the state "activated" and "no clearance", I would say?). Really it seems to me that one state variable with  three states could be enough: Enabled / no clearance, Enabled / Clearance and Not Active. 

Also, nicely, this way the airspace file does not need information about whether an airspace is "positive" or "negative", nor does it need to be derived e.g. from airspace class - the "Clearance" mechanism takes care of that property.

In terms of the altitude situation it might be necessary to store separate altitude information for the activation and the clearance though, as there are situations where a variable airspace might be active and you get additional clearances for other altitudes.

Example: There is a glider sector WHISKEY inside an otherwise restricted airspace, which is activated to varying altitudes or not at all on a given day.

Let's say it is active up to 10000 feet on this day, you enter it, and then ask to climb higher. A typical permission by the controllers is "Cleared to FL130 as long as you stay inside WHISKEY" (horizontally). So, ideally the system allows to both store an activation to 10000 MSL for the full day, but also the current clearance to FL130, but that clearance could then automatically expire once you leave the sector using the "ONCE" timer, or until you manually revoke it.

Wording:
======
I think it is best to use "ENABLED" and "DISABLED" rather than "ACTIVATED" / "NOT ACTIVATED", because it better describes how the software handles this, and to void confusions between what happens with "positive" relative to  "negative" airspaces- but this is open to discussion.

User interface: 
=========

All of this of course needs a more complex interface than the current "ACK" and "ACK DAY" buttons. There are many ways how to implement this. it COULD be like this:

Replace "ACK" / "ACK DAY" with four options:

"CLEAR ONCE", "CLEAR DAY", "DISABLE DAY". "MORE"

Then, MORE could open a dialog that allows to generate clearances / enabling / disabling with additional options.

That dialog would have fields for
"UP TO ALTITUDE"
and
"ABOVE ALTITUDE"
for setting the modified altitude limits for the clearance / activation,

(potentially two separate sets of fields to display/set the activation and clearance altitude limits separately, or some other solution  many possibilities here).

For the altitudes, there should be the option to set UNK, GND, XX00 FT MSL, and FLXXX values.

Also, at least on devices with big screens it would be super useful to have a couple of quick select buttons for typical values such as "above 9500MSL", "below FL100", "below FL195" etc. (could be hardcoded, though they might be different from country to country).

Then there should be the time limit options such as
"VALID UNTIL", with options such as
"ONCE", "ALL DAY", "NOW", or "5 MINUTES" or even an adjustable time in Minutes.

Then there are the buttons for executing, I am thinking of these functions (names can be different):

"DISABLE AIRSPACE"

"UPDATE and ENABLE AIRSPACE" (which keeps the airspace enabled, but changes altitude limits up until the specified time)

"REGISTER CLEARANCE" (which stores /updates the clearance until it expires / is deactivated)

"REVOKE CLEARANCE" removes the current clearance immediately

"CANCEL"

Conclusion
=======

This should cover the vast majority, if not all of the airspace situations. Please comment if something important is missing; or if there is a better / simpler solution with similar capabilities. And of course, if you know how to code this, help doing it or make suggestions on how it could best be done!

Of course, the features mentioned here can be implemented separately and sequentially rather than in one big update of the airspace system. I just hink it is best to discuss them together, such that any addition to the code is written in such a ways as to not make the other features harder to implement.

**Comment by kobedegeest** [comment]
> Acknowledgements: Currently, when entering an airspace which is not prohibited (either not active or due to ATC clearance), there are only two options to proceed: "Acknowledge", and "Acknowledge Day". The first makes the warning go away for a very short time before it returns, the other disables the warning for the entire day. In practice, often neither of these are what you want - e.g. you want to cross the airspace once, and then it should ideally be automatically reactivated (see below)

This acknowledge time can be set up to 16min40sec. But the easiest for this situation looks like just do "ack day" and when you say goodbye to ATC go to warnings en enable the airspace again. Using the input event system this dialog can be made easily accessible. It list all current warnings and all ack day airspaces.

> Clearance altitudes: When cleared by ATC into an airspace, often altitude limits are imposed. The controller will clear you into the airspace "up to flight level 100" or so, Or, a restriction area might be activated on a given day, but it sometimes is a different altitude up to which it is active on different days. I encounter both of these cases very often. So, there should be a way to not completely the warnings for a given airspace, but rather to (temporarily) adjust its altitude limits.

So what you need here is to "ack day" the airspace and activated some temporary altitude warning. Altitude warnings should be doable using lua. But also your in contact with ATC at this point so they will warn you as well.

> "Positive" airspaces: There are many airspaces which are not restrictions, but rather excepetions from restrictions of the bigger restrictive airspace they are inside of (when active): glider sectors and wave sectors, typically. Currently these either generate warnings, or are ignored; but really they should "disable" warnings as long as one is inside the sector (basically an inverted behaviour).

So the way we get around this is by just not having two overlapping airspaces for this. For the airspace that has a part that can be declassified we split the airspace in two parts (or more). The part that can get declassified and the part that does not get declassified. 

Overall i think that adding more buttons and on the fly settings is not an improvement. The "ack" and "ack day" buttons are already to small and hard to press. And every second you are busy pressing buttons just increase the chance of entering an airspace.

**Comment by groundhogxc** [comment]
> This acknowledge time can be set up to 16min40sec. But the easiest for this situation looks like just do "ack day" and when you say goodbye to ATC go to warnings en enable the airspace again. Using the input event system this dialog can be made easily accessible. It list all current warnings and all ack day airspaces.

This is what I currently do of course in the particular case of getting one clearance for one (restrictive) airspace. But keep in mind you are not necessarily in active contact with ATC after confirmation that some sector is active for example, So it is easily possible to forget, and anyway it can take mental work off the pilot.

> > Clearance altitudes: When cleared by ATC into an airspace, often altitude limits are imposed. The controller will clear you into the airspace "up to flight level 100" or so, Or, a restriction area might be activated on a given day, but it sometimes is a different altitude up to which it is active on different days. I encounter both of these cases very often. So, there should be a way to not completely the warnings for a given airspace, but rather to (temporarily) adjust its altitude limits.
> 
> So what you need here is to "ack day" the airspace and activated some temporary altitude warning. Altitude warnings should be doable using lua. But also your in contact with ATC at this point so they will warn you as well.

Yes, but then I would have to both reset the airspace manually, and reset the altitude warning.
And again, you might not be in active contact with ATC, you might ask "what is available altitude to day in this airspace zone", they give a number for the day.
And even if in contact, without a transponder they can not warn you at all.

But also, beyond the standard case of an explicit permission to enter some airspace (up to altidute XXX). A very typical (very common occurrnce for me): you might have a NOTAM that say something like "restricted area active from 1200 local to 1800 local". You cross the restricted zone at 1100 local in the morning, without any contact to ATC whatsoever. You disable the warning. Now it is easy to forget to re-activate it again after you left it, resulting in no warning when you come back home in the afternoon. With "disable once" this could also be fixed without entering a time like "disable for one hour" (although both could be reasonable solutions here - just do not press the "disable one hour" button at 11:30...).

We could also introduce "hours of activity" that can be set before the flight, but that is an additional, and also quite involved change.

> > "Positive" airspaces: [...]
> So the way we get around this is by just not having two overlapping airspaces for this. For the airspace that has a part that can be declassified we split the airspace in two parts (or more). The part that can get declassified and the part that does not get declassified.

I do not think that would be a good and simple implementation. You would have to do some maths to create new polygons, which is quite complicated - for example because the points of one airspace might not be exactly on the line of the other airspace, so you have to deal with the artifacts created.

Also, as you point out, airspaces might have to be split into two for each "positive" airspace depending on the geometry, and in case of one contained in the other at the very least an internal boundary will have to be created. It is all quite messy, and in order to not introduce a lot of additional relational information it would probably mean having to re-read the airspace file(s) after every change.

I think the most efficient way to implement this is where the warnings are generated: The algorithm searches for the nearest "problematic" airspace in the airspace database. That could be extended to search through the "Positive" airspaces as well, and then disable the warning if there is an intersection with a "positive" airspace within, say, 50m or 100m of the intersection with the problematic airspace. (some tolerance is required because of the points-on-the-line issue mentioned above). So no new algorithm, and very little additional computation - not least because the number of airspaces with "permission" turned on will always be much smaller than the total number of airspaces in the database.


> Overall i think that adding more buttons and on the fly settings is not an improvement. The "ack" and "ack day" buttons are already to small and hard to press. 

I agree more buttons should only be done when really useful, and here I think it is. But not much is added to the existing stuff. The main thing is an additional optional dialog.
Keep in mind that in the airspace warning dialog only one button is added (to access "more options"). So almost nothing changes if you do not want it to. Or, if you want (or on a very small screen), the "Ack Day" could be replaced with the "more options" button, leading to exact same button number and size as right now.
(And of course "Acknowledge Day" would have to be renamed to "Disable Day" in my proposal, but it would still work the same)

That you say they are too small now: That is a separate issue that can be easily addressed. I think you are correct, at least on my screen they could/should be significantly bigger.

> And every second you are busy pressing buttons just increase the chance of entering an airspace.

That particular problem does not apply, I think - at least the way I use XCS. Because, after you have seen (not yet anything pressed) the warning, there should be no more danger of entering the airspace. I usually do not "Ack" it anyway until it is very clear where I can/should go.

**Comment by kobedegeest** [comment]
So first to clarify my comment on positive airspaces, this is something we do in our airspace files so nothing in XCSoar. Sure this might be some work the first time but really only needed once. So we have quite some parts airspaces of class C that in the weekend get declasified to class G. So we split the Class C airspace in several parts, parts that can get declassified and parts that don't and if a sector is free to use that day we just ACK day for that part. 

Editing your airspace files seems easier than handeling "positive airspace" in xcsoar's source code. (Also i do not think any other program uses a concept like "positive airspace") 

But if you know of some specific case where this seems not doable maybe you could share a picture of that airspace layout.
But there should be no case of allowed and disallowed airspace overlapping. You are either allowed there or not. (at a given point in time) 

You propose lots of big things in one issue. One thing you could start with is introducing a check for when plane leaves an airspace. You can have a look in AirspaceEnterMonitor.cpp. GCE_AIRSPACE_LEAVE is already a valid glide computer event but it is never triggered or used. But adding that seems easy enough. Once something like that is in place things like acc once are possible or a popup or sound alert when leaving an airspace can be triggered by the input event system.

Also have a look at the existing issues with airspace warnings #200  #950 as well as the thermal assistent being drawn on top of the warning making it impossible to press certain buttons.

**Reply by groundhogxc** [reply]
> You propose lots of big things in one issue. 

Yes, as I explained in the end of the initial post, I think it is best to discuss the whole thing first, before the potentially breaking it up into sub-issues, such that the implementations do not interfere with one another - in particular in terms of what changes to data structures are the best choice.

**Reply by groundhogxc** [reply]
> So first to clarify my comment on positive airspaces, this is something we do in our airspace files so nothing in XCSoar. Sure this might be some work the first time but really only needed once. So we have quite some parts airspaces of class C that in the weekend get declasified to class G. So we split the Class C airspace in several parts, parts that can get declassified and parts that don't and if a sector is free to use that day we just ACK day for that part.

OK, I misunderstood you were proposing for XCSoar to do this automatically on loading the airspace files. I see that you manually do this.

Sure, this works, but it still has some of the drawbacks I mentioned (additional airspace borders or airspaces split into different parts), and, most of all, that is a TON of work to do (and lots of opportunities to make a mistake), for every country you fly in -- sometimes several times per year when new airspace files come out. In central Europe, countries are small, and have loads of airspaces, so, many files with many airspaces are needed :-)

I guess someone could do this centrally, and the modified airspaces could be on the repository...

> Editing your airspace files seems easier than handeling "positive airspace" in xcsoar's source code. (Also i do not think any other program uses a concept like "positive airspace")

Yes, I think no other program can do it so far - so that would be a real killer feature of XCSoar that sets it apart from the others! :-) And as I described, I think it is not functionally that complicated (to actually do it, to actually write the code, in the complex XCS code with all the abstraction layers and data structures however is not actually easy at all - my own level of coding skill is quite possibly inadequate).

But maybe my proposed method does not work because I overlooked something?

> But if you know of some specific case where this seems not doable maybe you could share a picture of that airspace layout. But there should be no case of allowed and disallowed airspace overlapping. You are either allowed there or not. (at a given point in time)

Of course it is doable by editing the "official" files. But in the "official" files such overlaps exist in many, many places. Most "wave" airspaces (which usually means "glider sector" or something similar) are exactly like this.
If you wanted to implement such a wave sector inside, say, a C airspace, with purely "traditional" airspace handling, you have to:

1) Modify the C airspace to have a "hole" where the wave sector is. I think currently XCS can not handle airspaces with holes, so you either have to:
1a) Run a "channel" from the edge of the airspace to the hole, or
1b) Split the airspace into two parts

2) Create a new C airspace with the shape of the wave sector, but above the wave sector 

3)  Create  a second new C airspace with the shape of the wave sector, with the lower edge according to the original C airspace, and the upper edge according to the wave sector (assuming the wave sector is not higher than the C airspace).

I assume this is what you are doing in the text file?

Now, when you cross these airspaces, you need to acknowledge the wave, or all of them individually when you get a C clearance, and then re-enable them all individually afterwards. 

(Of course, this assumes that the wave sector is always cleared to the same altitude, which is often not the case. In that case you would need several C airspace variants (if you wanted to have that functionality).)

I have thought about doing this in the past, but my current airspace file has more than 100 such sectors, many of whom intersect more than one airspace. That is just too much work for a lazy person like me, the files are updated too often, and the danger is big to make a mistake. It might also start to affect performance of XCS to add so many additional polygons (but maybe not an issue).

> You propose lots of big things in one issue. One thing you could start with is introducing a check for when plane leaves an airspace. 
Yes. That would be required to implement the "disable once" / "clearance once" feature - but with the current system it is not useful, as it would only work on active airspaces - which are generating a message anyway every few seconds/minutes, so reminding maybe not necessary - I guess everyone uses "Ack day" anyway. I think that is why it was never used / implemented so far.

> Also have a look at the existing issues with airspace warnings #200 #950 as well as the thermal assistent being drawn on top of the warning making it impossible to press certain buttons.

Thanks for the pointers, those are also relevant issues, yes. I just rarely use the cross section, and never the Radar, so this never affected me so far.

**Comment by lordfolken** [comment]
> You propose lots of big things in one issue.

I converted this to a discussion.
Please break out induvidual issues and refine them.

**Reply by groundhogxc** [reply]
Ah, OK. I am not yet used to this new "discussion" feature in github. Probably the right thing todo. Maybe you could give the original issue the "feature request" status, so that people can find it more easily (and thus find this discussion), if they have similar ideas and use the search on all issues. Maybe not important though.

**Comment by groundhogxc** [comment]
Let me clarify what I hope to achieve here:

- Have other people check the idea, whether it can work and makes sense (no internal contradiction)
- Gather comments about the methods, in particular whether the algorithm for positive airspace makes sense
- Gather comments about the additional data that needs to be handled in the airspace database (additional flages/states/altitude variables)
- Motivate people to help with implementation :-)
- Document the proposed feature for future reference, in case we can not do it now, but someone thinks about it later on
- Document the proposed feature so that, if someone works on other aspects of the airspace system, it can potentially be taken into account.

In particular, I am wondering about the "positive airspace" algorithm: Is it enough to extend the airspace check in the way described?
And also: What is the best set of states?

Currently I am thinking: Extending to three states:
"Disabled", "Enabled, no clearance", "Enabled, clearance".

Where "disabled" and "enabled, clearance" can have a time or the "once" flag associated with it, which reverts the state to "enabled" after the condition (time up, or leaving airspace, or manually revoking clearance) is met.

In the "variable altitude permission" extension, there would be a modified top altitude, a modified bottom altitude, or both, associated with the clearance, if chosen by the user. This would reset the permission to the normal values after the clearance has expired.

In the "variable altitude airspace" extension, there would be an additional top and/or bottom altitude associated with the airspace. This would be more like for air spaces that have the same altitude all day - say, an airspace that starts at one altitude on weekdays, and at another on weekends. Or a restricted area that goes very high on the map, but is activated by NOTAM or other publication to much lower altitudes, which are different each day (I hate those).

And whatever else you folks might come up with to improve the proposal.

**Comment by groundhogxc** [comment]
As suggested by @lordfolken, more specific implementation discussions for individual compoonents of this discussion are started in issues / feature requests #2375 and #2376.
